### ISSUE TITLE
Defaults need to be more mobile-friendly

### ISSUE THREAD

**Issue by jnm2** [jnm2/epic#6]
There is a lot of lag on my Android while drawing, resulting in a low-poly path. The FFT size should either be smaller or should adapt somehow.

**Comment by Daynvheur** [comment]
Fftsize can be forced to some value with the (not UI-fied) `fftsize=` url param.
Android lag is lessen since touch addPoint changes.

An input element also looks like a good idea.

**Comment by jnm2** [comment]
You're right, the lag seems to be gone. I'd still like to make it work well by default when circles are being drawn without having to first think about and set an FFT size setting.

**Comment by Daynvheur** [comment]
Late reply to review @jnm2:
- As paths only change when points are added, it is possible to prepare a background buffered path image easing the redraw() load
- As each circle+ray is a downscale (+repositionned+rotated) of their higher orders, a single circle+ray image may be used then downscaled if the cost of such operation has benefits over drawing circles and lines themselves (though it will degrade image sharpness)
- As each step has a finite result, those can be stored as background buffered images to avoid redraw() on already drawn step (at the cost of widly increasing memory footprint)
- FFT reduction is still something to consider, as there might be no use for 4096 subdivisions of a 16 points drawing, maybe starting at 128 and having an up/down input going for powers of 2 is an idea

**Comment by Daynvheur** [comment]
I'm heading for a PR with those:
- As closedPath only changes when a point is added, its computation is stored+reused
- As complexityPath+circlePath+linePath cycle through limited position defined by the parameter, their computation is stored in dictionaries<parameter, path> (reset when a point is added)
_^ This does reduce the render time and lighten cpu load, tho increasing memory footprint_
- The heaviest path is actually the circlePath, still in WIP stage for path improvement
- Fft is modifiable in an input field stepping by powers of 2 (ranging from 2 to 4096)
_Typical value of 1024 or even 512 may be enough as default values, there's room for discussion here_

**Comment by Daynvheur** [comment]
Got the AutoFFT working in the PR, yay!
Typical FFT value with 50k complexity is 1024, that's the default for mobile-detected browsers.

It will adapt to more (until 65536) up to much less (down to 2). Waiting for your review and user experience on this.

**Comment by Daynvheur** [comment]
Since the PR #9 has shipped automatic FFT adaptation, do you still have lags on android hardware?

**Comment by jnm2** [comment]
I'm not currently getting lag on android. There are separate bugs though.

**Comment by jnm2** [comment]
(https://github.com/jnm2/epic/issues/14 and https://github.com/jnm2/epic/issues/15)
